File: queue/queue.cpp
#include<iostream>

using namespace std;


template <typename T>
class Queue {
public:
    Queue(int c = 10);
    ~Queue();
    bool enqueue(T data);
    T dequeue();
    T front();
    bool isEmpty();
    bool isFull();
    int size();
    void info();
private:
    T *queue_;
    int front_;
    int rear_;
    int capacity_;
};

int main() {
    Queue<int> q;
    for (int i = 0; i < 11; i++) {
        q.enqueue(i);
        cout << q.size() << endl;
    }

    for (int j = 0; j < 11; j++) {
        if (!q.isEmpty())
            cout << q.dequeue() << endl;
        else
            break;
    }

    q.info();
    return 0;
}

template<typename T>
Queue<T>::Queue(int c):
    capacity_(c + 1),
    front_(0),
    rear_(0),
    queue_(nullptr) {
        queue_ = new T[c];
        cout << "capacity:" << c << endl;
    }

template <typename T>
Queue<T>::~Queue() {
    if(queue_) {
        delete [] queue_;
    }
}

template <typename T>
bool Queue<T>::isEmpty() {
    return (front_ == rear_);
}

template <typename T>
bool Queue<T>::isFull() {
    return (front_ == ((rear_ + 1) % capacity_));
}

template <typename T>
bool Queue<T>::enqueue(T data) {
    if (isFull()) {
        cout << "current data: " << data << ", queue is full" << endl;
        return false;
    };
    *(queue_ + rear_) = data;
    rear_ = (rear_ + 1) % capacity_;
    return true;
}

template <typename T>
T Queue<T>::dequeue() {
    if (isEmpty()) {
        throw out_of_range("Empty queue!");
    }
    T data = *(queue_ + front_);
    front_ = (front_ + 1) % capacity_;
    return data;
}

template <typename T> 
void Queue<T>::info() {
    cout << "front: " << front_ << ", tail: " << rear_ << endl;
}

template <typename T>
int Queue<T>::size() {
    return (rear_ - front_) % capacity_;
}

File: queue/queue.py
class Queue(object):
    """FIFO queue in python."""

    def __init__(self, maxsize):
        self.maxsize = maxsize
        self.queue = list()

    def push_back(self, x):
        """Push an element at the back of the queue.
        @Params:
            x: any type.
        @Return:
            bool,
                True if successful else False.
        """
        if not self.is_full():
            self.queue.insert(0, x)
            return True
        return False

    def pop(self):
        """Pop the front element of the queue.
        @Return:
            the front element if the queue is not empty,
            else raise a Error
        """
        if not self.is_empty():
            return self.queue.pop()
        else:
            print("Queue is empty!")
            exit()
    
    def is_full(self):
        """Check if the queue is full.
        @Return:
            True if the queue is full else False
        """
        if(len(self.queue) == self.maxsize):
            return True
        return False

    def is_empty(self):
        """Check if the queue is empty
        @Return:
            True if the queue is empty else False.
        """
        if len(self.queue) == 0:
            return True
        return False


if __name__ == '__main__':
    q = Queue(10)
    for i in range(11):
        if q.push_back(i):
            print("enqueue: {}".format(i))
        else:
            print("failed: {}".format(i))

    for i in range(11):
        if not q.is_empty():
            print("dequeue: {}".format(q.pop()))
        else:
            q.pop()
